File: packages/dart/lib/src/catalogs/banxico/tiie_28.dart
/// TIIE 28 días Catalog
///
/// Catalog of TIIE 28-day values from Banco de México
library;

/// TIIE 28-day Catalog
class Tiie28Catalog {
  static List<Map<String, dynamic>>? _data;

  /// Loads the TIIE data
  static void _loadData() {
    if (_data != null) return;
    // TODO: Load from packages/shared-data/banxico/tiie_28.json
    _data = [];
  }

  /// Gets all TIIE data
  static List<Map<String, dynamic>> getData() {
    _loadData();
    return List.from(_data!);
  }

  /// Gets most recent TIIE rate
  static Map<String, dynamic>? getActual() {
    _loadData();
    if (_data!.isEmpty) return null;
    final sorted = List.from(_data!)
      ..sort((a, b) => (b['fecha'] as String).compareTo(a['fecha'] as String));
    return sorted.first;
  }

  /// Gets TIIE rate for a specific date
  static Map<String, dynamic>? getPorFecha(String fecha) {
    _loadData();
    return _data!.firstWhere(
      (record) => record['fecha'] == fecha,
      orElse: () => <String, dynamic>{},
    );
  }

  /// Gets current TIIE rate value
  static double? getTasaActual() {
    final record = getActual();
    return record?['tasa'] as double?;
  }
}

File: packages/typescript/src/catalogs/banxico/tiie-28.ts
/**
 * TIIE 28 días Catalog
 *
 * Provides access to TIIE 28-day values from Banco de México.
 * TIIE (Tasa de Interés Interbancaria de Equilibrio) is the interbank equilibrium interest rate.
 */

import tiieData from '../../../../shared-data/banxico/tiie_28.json';

export interface TIIE28Record {
  fecha: string;
  tasa: number;
  plazo_dias: number;
  tipo: string;
  año: number;
  mes: number;
  fuente: string;
}

export class TIIE28Catalog {
  private static data: TIIE28Record[] = tiieData as TIIE28Record[];
  private static byFecha: Map<string, TIIE28Record> | null = null;

  private static _loadIndexes(): void {
    if (this.byFecha !== null) return;
    this.byFecha = new Map();
    for (const record of this.data) {
      this.byFecha.set(record.fecha, record);
    }
  }

  static getData(): TIIE28Record[] {
    return [...this.data];
  }

  static getPorFecha(fecha: string): TIIE28Record | null {
    this._loadIndexes();
    return this.byFecha!.get(fecha) ?? null;
  }

  static getActual(): TIIE28Record | null {
    if (this.data.length === 0) return null;
    return this.data.reduce((latest, record) => (record.fecha > latest.fecha ? record : latest));
  }

  static getTasaActual(): number | null {
    return this.getActual()?.tasa ?? null;
  }
}

export function getTIIEActual(): TIIE28Record | null {
  return TIIE28Catalog.getActual();
}
